docs/static/field-reference.asciidoc

```diff
@@ -0,0 +1,122 @@
+[[field-references]]
+== Field References
+
+It is often useful to be able to refer to a field or collection of fields by name. To do this,
+you can use the Logstash field reference syntax.
+
+The syntax to access a field specifies the entire path to the field, with each fragment wrapped in square brackets.
+
+_Field References_ can be expressed literally within <<conditionals,_Conditional_>> statements in your pipeline configurations,
+as string arguments to your pipeline plugins, or within sprintf statements that will be used by your pipeline plugins:
+
+[source,pipelineconf]
+filter {
+  #  +----literal----+     +----literal----+
+  #  |               |     |               |
+  if [@metadata][date] and [@metadata][time] {
+    mutate {
+      add_field {
+        "[@metadata][timestamp]" => "%{[@metadata][date]} %{[@metadata][time]}"
+      # |                      |    |  |               |    |               | |
+      # +----string-argument---+    |  +--field-ref----+    +--field-ref----+ |
+      #                             +-------- sprintf format string ----------+
+      }
+    }
+  }
+}
+
+[[formal-grammar]]
+=== Formal Grammar
```

Comment:
Who is the audience for this content? TBH, I find the details very hard to follow, maybe because I'm not familiar with the notation you're using here (antlr?) to represent the grammar. I'm concerned that users who aren't familiar with this style of notation will be confused by this topic. I don't think we can expect all of our users to have a programming background. If the goal is to help users understand how to use field references, I'd suggest providing more examples (maybe that show example event data and the syntax you would use to access specific fields). 

If a subset of users must know the formal grammar, then I'd consider adding it as an appendix rather than within the main flow of content. The way it's structured currently, the new topic breaks the book structure and doesn't fit logically into the hierarchy of topics.

Reply:
I agree that an appendix would be a useful place for this, the goal being that when someone runs into an issue where the loosey-goosey description of field references in other docs leads them to create inputs that fail to parse, they have somewhere to go to disambiguate while they debug.

How should I change this PR to make this an appendix entry?

***

I agree that the ANTLR-style grammar (or any formal meta-grammar) is a bit obtuse to an observer who is not familiar with formal grammar definitions, which is why each block is accompanied by an explanation in plain english using _Capital Emphasis_ to refer to proper nouns that are elsewhere defined in the document.

***

There may be value in re-arranging the sections to subtly get a reader into the habit of looking for more information in separate sections. 

 - The first section, _Field Reference Literal_, talks about _Path Fragments_, so it would be helpful if the _Path Fragment_ section immediately followed it.
 - The _Path Fragment_ section talks about _Field Names_, so it would be helpful if the _Field Name_ section immediately followed it.
  - A reader who gets this far will have much more success encountering the _Field Reference (Event APIs)_ section, which refers to _Field Reference Literal_  and  _Field Name_ (which they already understand), and _Composite Field Reference_ (which they don't know yet, and don't often _need_ to know, but should by now be familiar enough to guess that it is defined in this document should they need to drill in deeper).

Reply:
@yaauie To format this as an appendix, you need to move the topic to the end of the book  and tag the section correctly. Should look something like this:

```
["appendix",id="your-section-id"]
== Title

Your content
...

```